keps/sig-cli/4764-kubectl-watchlist/README.md

```diff
@@ -0,0 +1,743 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Update `kubectl` to support fetching collections using watch lists. This will allow the API server to serve
+those requests from its watch cache, which boosts performance and cuts the impact of those request
+on the API server.
+
+## Motivation
+
+Currently `client-go` could enable **watchlist** by env and it works great, but `kubectl` is still a pain. 
+And `kubectl` doest not have a cache mechanism like `client-go`, so it will do **list**  whenever it is used. Furthermore,
+request with pager params can not hit cache of `kube-apiserver`, every request will go through into `etcd`,
+which could push up the load of `kube-apiserver` and make system unstable.
+
+In practice, our users have been writing script to unrestrainedly call `kubectl`, which could lead serious performance
+issue. As cluster admininstor, we have few method to eliminate it. APF maybe is an option, but it's hard to configure
+and don't solve the problem from the root. In the other side, if we are standing in the shoes of user, most of those 
+scripts are simple Bash or Python which just simply glue a few `kubectl` commands. It is too heavy and "not worth" 
+for user to refactor just "a few commands" into "a batch of complex code". So we are seeking an optimization for `kubectl`. With this optimization, all users of `kubectl` can get free performance improvement.
+
+
+### Goals
+
+* Without breaking any user interface, add support to `kubectl` for fetching collections using **watch**.
+* Make `kubectl` hit the cache, thus reducing impact on `kube-apiserver`.
+
+### Non-Goals
+
+* Change `kubectl` functionality.
+
+## Proposal
+
+Existing `kubectl` subcommands like `get`, `describe` should be able to use a _watch list_ to fetch collections (list resources), rather than directly list resources page by page.
+
+At this stage, we will adopt a similar approach to `client-go`, using an environment variable to control whether
+to enable this experimental feature, and ensuring that it can revert to the previous behavior if it fails.
+
+<<[UNRESOLVED Will **watchlist** replace **list** completely]>>
+**watchlist** was first released in version 1.27 and is still in the alpha stage, with continuous improvements 
+over the subsequent releases. For a feature in such an early stage, it is too soon to say whether we will fully
+rely on it without other fallback mechanisms.
```

Comment:
kube-apiserver will support LIST requests forever for backward compatibility reasons.

But we expect that most clients (and libraries) will actually evolve towards using watchlist once this is graduated to GA.

Reply:
Yes - I think the things like `kubectl` should eventually migrate to WatchList